libyul/YulStack.h

```diff
@@ -163,6 +164,8 @@ class YulStack: public langutil::CharStreamProvider
 	std::shared_ptr<yul::Object> m_parserResult;
 	langutil::ErrorList m_errors;
 	langutil::ErrorReporter m_errorReporter;
+
+	std::shared_ptr<ObjectOptimizer> m_objectOptimizer;
```

Comment:
the instance is not shared around as far as i can tell so maybe a `unique_ptr` would be better

Reply:
It actually is. `CompilerStack` holds a master copy and passes it into various `YulStack`s on construction. So at any given moment it's held by both of them.

Technically, the object is only used in one place at a time, so I could still make it `unique_ptr` and move it in and out of `YulStack` but the design with `shared_pointer` seemed simpler to me.

Reply:
Ah interesting! I guess it doesn't matter much in this case then. Generally I find that `shared_ptr` are a bit overused at times where a `unique_ptr` and passing raw ptrs would do just as well - like here you could probably have a `unique_ptr` in the `CompilerStack` then. Or, you know, leave it as it is, which is fine too :D

Reply:
True, they are overused, we could pay more attention to that.

> and passing raw ptrs would do just as well - like here you could probably have a `unique_ptr` in the `CompilerStack` then.

But that happens only in `CompilerStack`. In other situations, e.g. for pure Yul compilation, `YulStack` does not get the cache from the outside and creates it on its own. It is then the only owner and has to use a smart ptr. So with a bare ptr I'd have to have a separate mechanism to handle that case.

I guess I could just always require the cache to be created outside and passed in. I didn't want to do that to keep `YulStack` self-contained.

On the other hand you could argue that always having to pass in the cache would be better, because this way you couldn't forget about it and unintentionally end up with a cache that is not shared even though it should be. But it also makes using `YulStack` more of a hassle and in many cases there's nothing to share the cache with anyway.

Reply:
Alright yeah, that makes sense. Thanks!